doc/reference/trd104-syscalls.md

```diff
@@ -596,12 +611,12 @@ The register arguments for Read-Write Allow system calls are as
 follows. The registers r0-r3 correspond to r0-r3 on CortexM and a0-a3
 on RISC-V.
 
-| Argument         | Register |
-|------------------|----------|
-| Driver number    | r0       |
-| Allow number     | r1       |
-| Address          | r2       |
-| Size             | r3       |
+| Argument         | Register | Type              | Value                                                                     |
+|------------------|----------|-------------------|---------------------------------------------------------------------------|
+| Driver number    | r0       | `u32`             |                                                                           |
+| Allow number     | r1       | `u32`             |                                                                           |
+| Address          | r2       | `*mut [u8; {r3}]` | Pointer to array of writable userspace memory of length `{r3}` **or** `0` |
+| Size             | r3       | `u32`             |                                                                           |
```

Comment:
One hitch here is that, actually, `ReadWriteAllow` uses `usize`, which is somewhat meaningful because it represents the length of an array which, in Rust and C and most languages is generally represented as word-sized (which makes sense since the word size generally limits the address space as well, so a 32-bit platform can't address an array larger than 4GB but a 64-bit platform can).

Reply:
Well yes and no. IIUC, the Rust usize conflated the C size_t and uintptr_t.

That was a fine enough thing to do because early Rust only targeted modern x86/64 and ARM, where that holds (note the modern qualifier is quite necessary there; e.g. MS-DoS's highmem is basically a form of bank switching or really anything that actually uses segments generally requires that objects fit in one segment but the pointer tells both which segment and where (size_t < ptr_t); more embedded-y example AVR today still uses segments with 24-bit addresses on 16-bit arch). And of course CHERI similarly exposes why one usize type can't represent both length and position.

I think my opposition to expressing anything in our syscall ABI in terms of usize comes from this observation that usize as-is is broken. The rust community is trying to figure out how to reckon with that reality. We don't have to.

As C types are sort of the lingua franca for FFI still, it probably would be better to use them in the Type column here; or perhaps to just have explicit Rust type and C type columns...